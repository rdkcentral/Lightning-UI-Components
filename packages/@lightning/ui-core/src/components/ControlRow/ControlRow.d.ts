import type lng from '@lightningjs/core';
import type { StylePartial } from '../../types/lui';
import TitleRow, { TitleRowStyle } from '../TitleRow';

export type ControlRowStyle = TitleRowStyle & {
  extraItemSpacing: number;
};

type ControlRowItem = Array<
  lng.Component.NewPatchTemplate<typeof lng.Component> | lng.Component
>;

export default class ControlRow extends TitleRow {
  leftControls?: ControlRowItem;
  contentItems?: ControlRowItem;
  rightControls?: ControlRowItem;
  extraItemSpacing?: number;
  get style(): ControlRowStyle;
  set style(v: StylePartial<ControlRowStyle>);

  loadMoreItems(): void;
  addContentItems(items: ControlRowItem[]);
  addContentItemsAt(items: ControlRowItem[], index: number);
  removeContentItemsAt(index: number);
  addLeftControls(items: ControlRowItem[]);
  addLeftControlsAt(items: ControlRowItem[], index: number);
  removeLeftControlsAt(index: number);
  addRightControls(items: ControlRowItem[]);
  addRightControlsAt(items: ControlRowItem[], index: number);
  removeRightControlsAt(index: number);
}
