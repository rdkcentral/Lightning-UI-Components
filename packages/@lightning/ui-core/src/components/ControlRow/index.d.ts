import ControlRow, { ControlRowStyles } from './ControlRow';

export { ControlRow as default, ControlRowStyles };
