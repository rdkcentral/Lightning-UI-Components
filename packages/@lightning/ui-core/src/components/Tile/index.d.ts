import Tile, { TileStyle } from './Tile';

export { Tile as default, TileStyle };
