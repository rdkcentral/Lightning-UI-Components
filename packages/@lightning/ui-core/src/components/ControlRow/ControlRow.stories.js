import lng from '@lightningjs/core';
import Tile from '../Tile';
import ControlSmall from '../Control/ControlSmall';
import lightningIcon from '../../assets/images/ic_lightning_white_32.png';
import playIcon from '../../assets/images/ic_play_white_32.png';
import { default as TitleRowStory } from '../TitleRow/TitleRow.stories';
import ControlRowComponent from '.';
import mdx from './ControlRow.mdx';
import { CATEGORIES } from 'lightning-ui-docs';
import { createModeControl } from 'lightning-ui-docs/.storybook/utils';

export default {
  title: `${CATEGORIES[64]}/ControlRow`,
  parameters: {
    docs: {
      page: mdx
    }
  },
  args: {
    ...TitleRowStory.args,
    lazyLoadBuffer: 0,
    lazyScroll: true,
    mode: 'focused'
  },
  argTypes: {
    ...TitleRowStory.argTypes,
    ...createModeControl({ summaryValue: 'focused' }),
    lazyLoadBuffer: {
      control: 'number',
      description:
        'Number of content items prior to last content items where a "loadMoreItems" signal is emitted. Go to the "Actions" pannel to see when signal is fired.',
      table: { defaultValue: { summary: 0 } }
    },
    lazyScroll: {
      control: 'boolean',
      description:
        'Will only scroll the row if the item is off screen and alwaysScroll and neverScroll are both false',
      table: { defaultValue: { summary: true } }
    }
  }
};

const createItems = (length, src) => {
  return Array.from({ length }).map((_, index) => ({
    type: Tile,
    artwork: {
      src
    },
    itemLayout: {
      ratioX: 16,
      ratioY: 9,
      upCount: 4
    },
    announce: 'Tile',
    announceContext: `${index + 1} of ${length}`
  }));
};
const createControls = (length, icon = lightningIcon) => {
  return Array.from({ length }).map(() => ({
    type: ControlSmall,
    icon,
    fixed: true,
    h: 64,
    w: 64
  }));
};

export const Basic = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        ControlRow: {
          type: ControlRowComponent,
          leftControls: createControls(3),
          contentItems: createItems(5),
          rightControls: createControls(3)
        }
      };
    }

    _getFocused() {
      return this.tag('ControlRow');
    }
  };

export const LazyLoading = () =>
  class LazyLoading extends lng.Component {
    static _template() {
      return {
        ControlRow: {
          type: ControlRowComponent,
          leftControls: createControls(3),
          contentItems: createItems(5),
          rightControls: createControls(3),
          lazyLoadBuffer: 1,
          signals: {
            loadMoreItems: '_loadMore'
          }
        }
      };
    }

    _construct() {
      this._counter = 0;
    }

    _loadMore() {
      if (this._counter < 2) {
        this._ControlRow.addContentItems(
          createItems(
            3,
            'https://image.tmdb.org/t/p/w500/frwl2zBNAl5ZbFDJGoJv0mYo0rF.jpg'
          )
        );
        this._counter++;
      }
    }

    _getFocused() {
      return this.tag('ControlRow');
    }

    get _ControlRow() {
      return this.tag('ControlRow');
    }
  };

LazyLoading.args = { lazyLoadBuffer: 1 };

export const AddingAndRemoving = () =>
  class AddingAndRemoving extends lng.Component {
    static _template() {
      return {
        ControlRow: {
          type: ControlRowComponent,
          leftControls: createControls(1),
          contentItems: createItems(2),
          rightControls: createControls(1)
        }
      };
    }

    _construct() {
      super._construct();
      setTimeout(() => {
        this._ControlRow.addLeftControlsAt(createControls(1, playIcon), 0);
      }, 1500);
      setTimeout(() => {
        this._ControlRow.addContentItemsAt(
          createItems(
            2,
            'https://image.tmdb.org/t/p/w500/frwl2zBNAl5ZbFDJGoJv0mYo0rF.jpg'
          ),
          1
        );
      }, 3000);
      setTimeout(() => {
        this._ControlRow.addRightControls(createControls(3, playIcon));
      }, 4500);
      setTimeout(() => {
        this._ControlRow.removeRightControlAt(0);
      }, 6000);
    }

    _getFocused() {
      return this.tag('ControlRow');
    }

    get _ControlRow() {
      return this.tag('ControlRow');
    }
  };
