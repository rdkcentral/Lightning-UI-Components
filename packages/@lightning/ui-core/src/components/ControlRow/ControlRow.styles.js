export const base = theme => ({
  extraItemSpacing: theme.spacer.lg
});
