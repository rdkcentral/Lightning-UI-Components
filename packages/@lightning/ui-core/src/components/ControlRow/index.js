import ControlRow from './ControlRow';

export { ControlRow as default };
