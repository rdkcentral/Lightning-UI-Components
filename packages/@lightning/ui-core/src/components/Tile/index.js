export { default as default } from './Tile.js';
