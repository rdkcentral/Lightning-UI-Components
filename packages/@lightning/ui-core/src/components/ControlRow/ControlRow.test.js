import { makeCreateComponent } from '@lightning/ui-test-utils';
import Tile from '../Tile';
import lightningLogo from '../../assets/images/ic_lightning_white_32.png';
import ControlRow from '.';
import { jest } from '@jest/globals';

const createItem = (props = {}) => ({
  type: Tile,
  itemLayout: {
    ratioX: 16,
    ratioY: 9,
    upCount: 4
  },
  ...props
});

const createItems = length => {
  return Array.from({ length }).map(() => createItem());
};

const createControl = (props = {}) => ({
  logo: lightningLogo,
  ...props
});

const createControls = length => {
  return Array.from({ length }).map(() => createControl());
};

const leftControls = createControls(2);
const rightControls = createControls(2);
const items = createItems(2);

const createComponent = makeCreateComponent(ControlRow);

describe('ControlRow', () => {
  let component, testRenderer;

  beforeEach(() => {
    [component, testRenderer] = createComponent({
      contentItems: items
    });
  });

  it('renders', () => {
    [component, testRenderer] = createComponent({
      title: 'Title',
      leftControls,
      contentItems: items,
      rightControls
    });
    testRenderer.forceAllUpdates();
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  describe('updating the content in a row', () => {
    it('should add controls and items to the row', () => {
      [component, testRenderer] = createComponent({ leftControls });
      testRenderer.forceAllUpdates();
      expect(component.items.length).toBe(leftControls.length);
      expect(component._getLeftControlItems().length).toBe(leftControls.length);

      component.contentItems = items;
      testRenderer.forceAllUpdates();
      expect(component.items.length).toBe(leftControls.length + items.length);
      expect(component._getContentItemItems().length).toBe(items.length);

      component.rightControls = rightControls;
      testRenderer.forceAllUpdates();
      expect(component.items.length).toBe(
        leftControls.length + items.length + rightControls.length
      );
      expect(component._getRightControlItems().length).toBe(
        rightControls.length
      );
    });

    it('should add controls and items to the row via the ComponentList setter', () => {
      [component, testRenderer] = createComponent({});

      component.leftControls = leftControls;
      testRenderer.forceAllUpdates();
      expect(component.items.length).toBe(leftControls.length);
      expect(component._getLeftControlItems().length).toBe(leftControls.length);

      component.contentItems = items;
      testRenderer.forceAllUpdates();
      expect(component.items.length).toBe(leftControls.length + items.length);
      expect(component._getContentItemItems().length).toBe(items.length);

      component.rightControls = rightControls;
      testRenderer.forceAllUpdates();
      expect(component.items.length).toBe(
        leftControls.length + items.length + rightControls.length
      );
      expect(component._getRightControlItems().length).toBe(
        rightControls.length
      );
    });

    it('should add extra space between controls and and items', () => {
      component.leftControls = leftControls;
      component.contentItems = items;
      component.rightControls = rightControls;
      testRenderer.forceAllUpdates();
      const spacing = component.style.extraItemSpacing;
      expect(component.items[0].extraItemSpacing).toBeUndefined();
      expect(component.items[1].extraItemSpacing).toBe(spacing);
      expect(component.items[2].extraItemSpacing).toBeUndefined();
      expect(component.items[3].extraItemSpacing).toBe(spacing);
      expect(component.items[4].extraItemSpacing).toBeUndefined();
      expect(component.items[5].extraItemSpacing).toBeUndefined();
    });

    it('should allow overwriting extra item spacing space between controls and and items', () => {
      const spacing = 3;
      component.extraItemSpacing = spacing;
      component.leftControls = leftControls;
      component.contentItems = items;
      component.rightControls = rightControls;
      testRenderer.forceAllUpdates();
      expect(component.items[0].extraItemSpacing).toBeUndefined();
      expect(component.items[1].extraItemSpacing).toBe(spacing);
      expect(component.items[2].extraItemSpacing).toBeUndefined();
      expect(component.items[3].extraItemSpacing).toBe(spacing);
      expect(component.items[4].extraItemSpacing).toBeUndefined();
      expect(component.items[5].extraItemSpacing).toBeUndefined();
    });

    it('should default the selected item to the first item', () => {
      component.leftControls = leftControls;
      component.contentItems = items;
      component.rightControls = rightControls;
      testRenderer.forceAllUpdates();
      expect(component.selectedIndex).toBe(component.leftControls.length);
    });

    it('should allow setting the selected index', () => {
      component.leftControls = leftControls;
      component.contentItems = items;
      component.rightControls = rightControls;
      component.selectedIndex = 0;
      testRenderer.forceAllUpdates();
      expect(component.selectedIndex).toBe(0);
    });
  });

  describe('add - adding to ControlRow', () => {
    beforeEach(() => {
      component.leftControls = leftControls;
      component.contentItems = items;
      component.rightControls = rightControls;

      testRenderer.forceAllUpdates();
      expect(component.selectedIndex).toBe(2);
    });

    describe('left controls', () => {
      it('should persist the selected item when an item is added before the selected item index', () => {
        component.addLeftControls(leftControls);
        expect(component.selectedIndex).toBe(4);
      });
      it('should persist the selected item when an item is added at the selected item index', () => {
        component.selectedIndex = 1;
        component.addLeftControls(leftControls);
        expect(component.selectedIndex).toBe(1);
      });
      it('should persist the selected item when an item is added after the selected item index', () => {
        component.selectedIndex = 0;
        testRenderer.forceAllUpdates();
        component.addLeftControls(leftControls);
        expect(component.selectedIndex).toBe(0);
      });
    });

    describe('right controls', () => {
      it('should persist the selected item when an item is added before the selected item index', () => {
        component.addRightControls(rightControls);
        expect(component.selectedIndex).toBe(2);
      });
    });

    describe('content items', () => {
      it('should persist the selected item when an item is added at selected item index', () => {
        component.addContentItems(items);
        expect(component.selectedIndex).toBe(2);
      });
      it('should persist the selected item when an item is added after the selected item index', () => {
        component.selectedIndex = 1;
        component.addContentItems(items);
        expect(component.selectedIndex).toBe(1);
      });
    });
  });

  describe('addAt - adding at a specified index to ControlRow', () => {
    beforeEach(() => {
      component.leftControls = createControls(3);
      component.contentItems = createItems(3);
      component.rightControls = createControls(3);
      testRenderer.forceAllUpdates();
    });

    const controlId = 'newControl';
    const itemId = 'newItem';
    const newControl = createControl({ testId: controlId });
    const newItem = createItem({ testId: itemId });

    describe('left controls', () => {
      it('should add a control to the end of the left controls', () => {
        component.addLeftControlsAt([newControl], 3);
        expect(component.items[3].testId).toBe(controlId);
        expect(component.leftControls.length).toBe(4);
      });
      it('should add a control at a specific index', () => {
        component.addLeftControlsAt([newControl], 2);
        expect(component.items[2].testId).toBe(controlId);
        expect(component.leftControls.length).toBe(4);
      });
    });

    describe('right controls', () => {
      it('should add a control to the end of the right controls', () => {
        component.addRightControlsAt([newControl], 3);
        expect(component.items[9].testId).toBe(controlId);
        expect(component.rightControls.length).toBe(4);
      });
      it('should add a control at a specific index', () => {
        component.addRightControlsAt([newControl], 2);
        expect(component.items[8].testId).toBe(controlId);
        expect(component.rightControls.length).toBe(4);
      });
    });

    describe('content items', () => {
      let initialRowItemsLength;
      beforeEach(() => {
        initialRowItemsLength = component.items.length;
      });
      it('should add an item to the end of the items', () => {
        component.addContentItemsAt([newItem], 3);
        expect(component.items[6].testId).toBe(itemId);
        expect(component.contentItems.length).toBe(4);
        expect(component.items.length).toBe(initialRowItemsLength + 1);
      });
      it('should add an item at a specific index', () => {
        component.addContentItemsAt([newItem], 2);
        expect(component.items[5].testId).toBe(itemId);
        expect(component.contentItems.length).toBe(4);
        expect(component.items.length).toBe(initialRowItemsLength + 1);
      });
    });
  });

  describe('removeAt - removing an item or control from ControlRow', () => {
    beforeEach(() => {
      component.leftControls = createControls(3);
      component.contentItems = createItems(3);
      component.rightControls = createControls(3);
      testRenderer.forceAllUpdates();
    });

    describe('left controls', () => {
      it('should remove a control', () => {
        component.removeLeftControlAt(2);
        expect(component.leftControls.length).toBe(2);
      });
    });

    describe('right controls', () => {
      it('should remove a control', () => {
        component.removeRightControlAt(2);
        expect(component.rightControls.length).toBe(2);
      });
    });

    describe('content items', () => {
      let initialRowItemsLength;
      let initialContentItemsLength;
      beforeEach(() => {
        initialRowItemsLength = component.items.length;
        initialContentItemsLength = component._getContentItemItems().length;
      });
      it('should remove an item', () => {
        component.removeContentItemAt(2);
        expect(component._getContentItemItems().length).toBe(
          initialContentItemsLength - 1
        );
        expect(component.items.length).toBe(initialRowItemsLength - 1);
      });
    });
  });

  describe('signaling to the consumer to load more items', () => {
    let signalSpy;
    beforeEach(() => {
      jest.clearAllMocks();
      signalSpy = jest.spyOn(component, 'signal');
      component.contentItems = createItems(6);
      component.lazyLoadBuffer = 2;
    });
    it('should signal if an item is selected at or after the load more index', () => {
      component.selectedIndex = 1;
      expect(signalSpy).not.toHaveBeenCalledWith('loadMoreItems');

      component.selectNext();
      expect(signalSpy).not.toHaveBeenCalledWith('loadMoreItems');

      component.selectNext();
      expect(signalSpy).toHaveBeenCalledWith('loadMoreItems');
      signalSpy.mockClear();

      expect(signalSpy).not.toHaveBeenCalledWith('loadMoreItems');
      component.selectNext();
      expect(signalSpy).toHaveBeenCalledWith('loadMoreItems');
    });

    it('should not signal when a control is selected', () => {
      component.selectedIndex = 0;
      expect(signalSpy).not.toHaveBeenCalledWith('loadMoreItems');

      component.selectPrevious();
      expect(signalSpy).not.toHaveBeenCalledWith('loadMoreItems');

      component.selectedIndex = 6;
      component.selectNext();
      expect(signalSpy).not.toHaveBeenCalledWith('loadMoreItems');
    });

    it('should not signal when a lazy scroll buffer has not been set', () => {
      component.lazyLoadBuffer = undefined;
      component.selectedIndex = 3;
      expect(signalSpy).not.toHaveBeenCalledWith('loadMoreItems');

      component.selectNext();
      expect(signalSpy).not.toHaveBeenCalledWith('loadMoreItems');
    });
  });
});
