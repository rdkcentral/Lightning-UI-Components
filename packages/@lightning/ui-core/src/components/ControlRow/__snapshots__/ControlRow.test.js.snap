// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlRow renders 1`] = `
{
  "Component": {
    "Items": {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Element-37": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 108.5,
          "zIndex": 0,
        },
        "Element-38": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "visible": true,
          "w": 0,
          "x": 24,
          "y": 108.5,
          "zIndex": 0,
        },
        "Element-39": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Background": {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Background",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Background",
              ],
              "texture": {
                "type": "StaticCanvasTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
            "Tile": {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "children": {
                "Artwork": {
                  "active": true,
                  "alpha": 0.001,
                  "attached": true,
                  "boundsMargin": null,
                  "children": {
                    "Image": {
                      "active": true,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 0,
                      "isComponent": undefined,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": "Image",
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": undefined,
                      "tag": [Function],
                      "tags": [
                        "Image",
                      ],
                      "visible": true,
                      "w": 0,
                      "x": 0,
                      "y": 0,
                      "zIndex": 0,
                    },
                  },
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 217,
                  "hasFinalFocus": false,
                  "hasFocus": false,
                  "isComponent": true,
                  "mount": 0.5,
                  "mountX": 0.5,
                  "mountY": 0.5,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Artwork",
                  "renderOfScreen": undefined,
                  "renderToTexture": true,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": "",
                  "tag": [Function],
                  "tags": [
                    "Artwork",
                  ],
                  "type": "Artwork",
                  "visible": true,
                  "w": 386,
                  "x": 193,
                  "y": 108.5,
                  "zIndex": 0,
                },
                "Content": {
                  "active": true,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 217,
                  "isComponent": undefined,
                  "mount": 0.5,
                  "mountX": 0.5,
                  "mountY": 0.5,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Content",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Content",
                  ],
                  "visible": true,
                  "w": 386,
                  "x": 193,
                  "y": 108.5,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Tile",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Tile",
              ],
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 217,
          "hasFinalFocus": true,
          "hasFocus": true,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Tile",
          "visible": true,
          "w": 386,
          "x": 48,
          "y": 0,
          "zIndex": 0,
        },
        "Element-45": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Background": {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Background",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Background",
              ],
              "texture": {
                "type": "StaticCanvasTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
            "Tile": {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "children": {
                "Artwork": {
                  "active": true,
                  "alpha": 0.001,
                  "attached": true,
                  "boundsMargin": null,
                  "children": {
                    "Image": {
                      "active": true,
                      "alpha": 1,
                      "attached": true,
                      "boundsMargin": null,
                      "clipping": false,
                      "color": 4294967295,
                      "enabled": true,
                      "flex": false,
                      "flexItem": false,
                      "h": 0,
                      "isComponent": undefined,
                      "mount": 0,
                      "mountX": 0,
                      "mountY": 0,
                      "pivot": 0.5,
                      "pivotX": 0.5,
                      "pivotY": 0.5,
                      "ref": "Image",
                      "renderOfScreen": undefined,
                      "renderToTexture": false,
                      "scale": 1,
                      "scaleX": 1,
                      "scaleY": 1,
                      "state": undefined,
                      "tag": [Function],
                      "tags": [
                        "Image",
                      ],
                      "visible": true,
                      "w": 0,
                      "x": 0,
                      "y": 0,
                      "zIndex": 0,
                    },
                  },
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 217,
                  "hasFinalFocus": false,
                  "hasFocus": false,
                  "isComponent": true,
                  "mount": 0.5,
                  "mountX": 0.5,
                  "mountY": 0.5,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Artwork",
                  "renderOfScreen": undefined,
                  "renderToTexture": true,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": "",
                  "tag": [Function],
                  "tags": [
                    "Artwork",
                  ],
                  "type": "Artwork",
                  "visible": true,
                  "w": 386,
                  "x": 193,
                  "y": 108.5,
                  "zIndex": 0,
                },
                "Content": {
                  "active": true,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 217,
                  "isComponent": undefined,
                  "mount": 0.5,
                  "mountX": 0.5,
                  "mountY": 0.5,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Content",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Content",
                  ],
                  "visible": true,
                  "w": 386,
                  "x": 193,
                  "y": 108.5,
                  "zIndex": 0,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Tile",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": [
                "Tile",
              ],
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 217,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "type": "Tile",
          "visible": true,
          "w": 386,
          "x": 458,
          "y": 0,
          "zIndex": 0,
        },
        "Element-51": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "visible": true,
          "w": 0,
          "x": 868,
          "y": 108.5,
          "zIndex": 0,
        },
        "Element-52": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": null,
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "visible": true,
          "w": 0,
          "x": 892,
          "y": 108.5,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 217,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Items",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Items",
      ],
      "visible": true,
      "w": 892,
      "x": -48,
      "y": 72,
      "zIndex": 0,
    },
    "Title": {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Text": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Text",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "Text",
          ],
          "texture": {
            "fontFace": "XfinityBrownBold",
            "fontSize": 36,
            "letterSpacing": -0.2,
            "lineHeight": 48,
            "maxLinesSuffix": "...",
            "text": "Title",
            "textBaseline": "bottom",
            "textColor": 4294375161,
            "textOverflow": "ellipsis",
            "type": "TextTexture",
            "verticalAlign": "middle",
          },
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 2,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 48,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Title",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": [
        "Title",
      ],
      "type": "TextBox",
      "visible": true,
      "w": 4,
      "x": 24,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
